src/main/java/com/mathworks/ci/tools/MatlabInstaller.java

```diff
@@ -0,0 +1,296 @@
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.Launcher.ProcStarter;
+
+import hudson.model.Node;
+import hudson.model.TaskListener;
+import hudson.tools.DownloadFromUrlInstaller;
+import hudson.tools.ToolInstallation;
+import hudson.tools.ToolInstallerDescriptor;
+
+import hudson.util.ArgumentListBuilder;
+import hudson.util.FormValidation;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.URL;
+
+import java.nio.charset.StandardCharsets;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Locale;
+
+import java.util.Set;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.IOUtils;
+
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.verb.POST;
+
+public class MatlabInstaller extends DownloadFromUrlInstaller {
+
+    private String version;
+    private String products;
+
+    @DataBoundConstructor
+    public MatlabInstaller (String id) {
+        super (id);
+    }
+
+    public String getVersion () {
+        return this.version;
+    }
+
+    @DataBoundSetter
+    public void setVersion (String version) {
+        this.version = version;
+    }
+
+    public String getProducts () {
+        return this.products;
+    }
+
+    @DataBoundSetter
+    public void setProducts (String products) {
+        this.products = products;
+    }
+
+    @Override
+    public FilePath performInstallation (ToolInstallation tool, Node node, TaskListener log)
+        throws IOException, InterruptedException {
+        FilePath expectedPath = preferredLocation (tool, node);
+        MatlabInstallable installable;
+        try {
+            installable = (MatlabInstallable) getInstallable (node);
+        } catch (Exception e) {
+            throw new InstallationFailedException (e.getMessage ());
+        }
+
+        getFreshCopyOfExecutables (installable, expectedPath);
+
+        FilePath versionInfo = new FilePath (expectedPath,"VersionInfo.xml");
+        FilePath installedProducts = new FilePath (expectedPath,"installed_matlab_product_list.txt");
+        if (versionInfo.exists () && isSameProduct(installedProducts)) {
+            return expectedPath;
+        } else {
+            int result = installUsingMpm (node, expectedPath, log, installedProducts);
+            if (result == 0) {
+                log.getLogger ().println (
+                    "MATLAB installation for version " + this.getVersion ()
+                        + " using mpm is completed successfully !");
+                updateProductList (installedProducts);
+            }
+        }
+        return expectedPath;
+    }
+
+    private int installUsingMpm (Node node, FilePath expectedPath, TaskListener log, FilePath installedProducts)
```

Comment:
No its not parsing the products, `FilePath installedProducts = new FilePath (expectedPath,"installed_matlab_product_list.txt");` just create a FilePath object and passes it 

i am later checking if this file exists that means we already have installed product list present. and make sure if the given product matches any of these lists to avoid invoking mpm 

if this does not exist that mean this is first mpm run we will then update the file with given products or at least with "MATLAB" if product list is empty .

Reply:
I think it would be helpful (make the code easier for us to maintain later) if the parameters were parsed by our jenkins plugin, and this `installUsingMpm` function operated very similarly to the `mpm install` API, which takes `destination`, `release` and `products` as arguments.

Reply:
In my opinion installUsingMpm needs the knowledge preintsalled products, However i think this may get removed if we decide not to maintain the product list. I will lchage it accordingly if we decide otherwise.